### PR TITLE
add unit tests for the case not allow to create server and fix missing server domain name in provision result

### PR DESCRIPTION
**Pull request by zhongyi-zhang**
…g server domain name in provision result

### CODE REVIEW

File: lib/services/azuresqldb/cmd-provision.js [hunk 1]
```diff
@@ -167,6 +167,7 @@ var sqldbProvision = function (log, params) {
                         var errorMessage = util.format('You specify the firewall rules for the server but you do not have the privilege to operate the server.');
                         callback(Error(errorMessage));
                     } else {
+                        fullyQualifiedDomainName = JSON.parse(result.body).properties.fullyQualifiedDomainName;
```
Comment (AbelHu) [inline lib/services/azuresqldb/cmd-provision.js]: Will it throw an error if result.statusCode !== HttpStatus.OK?

Comment (zhongyi-zhang) [inline lib/services/azuresqldb/cmd-provision.js]: Oh yes, the status code could be neither 404 nor 200. Will fix.

Comment (zhongyi-zhang) [inline lib/services/azuresqldb/cmd-provision.js]: @AbelHu Fixed.

Comment (AbelHu) [inline lib/services/azuresqldb/cmd-provision.js]: The fix looks good to me. You should add unit tests later for these two 'else'

File: lib/services/azuresqldb/client.js [hunk 2]
```diff
@@ -247,15 +248,13 @@ sqldbOperations.prototype.executeSql = function (config, sql, isDeleteFirewallRu
...
-      
+
             that.log.error('sqldb client: connect to server: err %j', err);
             return callback(err);
         }
 
+        that.log.info('sqldb: Connect to database %s in server %s: succeed.', config.database, config.server);
         that.log.info('sqldb: It is going to execute sql: %s', sql);
         var req = new sqlDb.Request(conn);
         req.query(sql, function(err) {
```
Comment (jinmiaoms) [inline lib/services/azuresqldb/client.js]: You don't check delete fire wall if sql execution fails?

File: lib/services/azuresqldb/cmd-provision.js [hunk 3]
```diff
@@ -166,8 +166,12 @@ var sqldbProvision = function (log, params) {
...
-                    } else {
+                    } else if (result.statusCode === HttpStatus.OK) {
+                        fullyQualifiedDomainName = JSON.parse(result.body).properties.fullyQualifiedDomainName;
                         callback(null);
+                    } else {
+                        var errorMessage = util.format('Unexpected error: %j', result);
+                        callback(Error(errorMessage));
                     }
                 },
                 function (callback) {   // see if db exists (in the case that sql server was there already)
```
Comment (bingosummer) [inline lib/services/azuresqldb/cmd-provision.js]: @zhongyi-zhang The integration test case is missing for the existing sql server.

File: lib/services/azuresqldb/cmd-provision.js [hunk 4]
```diff
@@ -230,10 +234,14 @@ var sqldbProvision = function (log, params) {
                                 callback(null);
                             }
                         });
-                    } else {  // sql server exists
+                    } else if (result.statusCode === HttpStatus.OK) {  // sql server exists
```
Comment (jinmiaoms) [inline lib/services/azuresqldb/cmd-provision.js]: Wouldn't the logic be cleaner (and saving a few lines of code) if we do getServer and check for OK status first?  That's always required anyway.  Only when getServer fails would we need to worry about if we have privilege to create server.

File: lib/services/azuresqldb/client.js (hunk 2 above)
Comment (zhongyi-zhang) [inline lib/services/azuresqldb/client.js]: Will fix.

File: lib/services/azuresqldb/cmd-provision.js (hunk 3 above)
Comment (zhongyi-zhang) [inline lib/services/azuresqldb/cmd-provision.js]: As we discuss before, the integration test case for the existing sql server can't be added easily. We can't change the privilege whether allow to create server or not. Though we have such a case, we still need to modify some env variables and the test cases to use. I don't think it is good for long term. We need to figure out a better way to do this.

File: lib/services/azuresqldb/cmd-provision.js (hunk 4 above)
Comment (zhongyi-zhang) [inline lib/services/azuresqldb/cmd-provision.js]: Thanks for the suggestion. Will fix.
The reason why the logic like this is, the broker doesn't have the option to set privilege whether allow to create server before. At that time, we assumed the server is usually a new one. So it checks if statusCode === 404 first.

File: lib/services/azuresqldb/client.js [hunk 5]
```diff
@@ -247,15 +248,13 @@ sqldbOperations.prototype.executeSql = function (config, sql, isDeleteFirewallRu
                         callback(result);
                     }
                 });
-            } else if (!err.message.startsWith('Login failed for user')) {
-                // It is for the scenario creating a new server, the master database is not ready at the same time the database created by the broker is ready.
-                return setTimeout(function(){that.executeSql(config, sql, isDeleteFirewallRuleAfter, callback);}, 1000);
             }
-
```
Comment (bingosummer) [inline lib/services/azuresqldb/client.js]: So we don't need this error checking?

**Pull request merged**